notebook: Add notebook estimate V100 GPU-days for cyclic adversary

tomtseng opened the pull request:
* Add notebook that parsers the cyclic adversary's victimplay logs to estimate that the cyclic adversary used 2,223 V100 GPU-days of training for victimplay (training GPU not included)

ed1d1a8d commented on notebooks/notebooks/iclr2022/estimate-gpu-days-adv.ipynb:
```diff
@@ -0,0 +1,321 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #15.    # - On cp505, 3 A6000 victimplay workers generated 250k rows in 9h30m and 7
</pre><p>Reminds of me of mathcounts problems haha.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AlignmentResearch/KataGoVisualizer/pull/62/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

ed1d1a8d replied:
<pre class="ql-syntax" spellcheck="false">Line #12.    gpu_type_to_seconds = {}
</pre><p>Could also use a defaultdict(int) to simplify code below.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AlignmentResearch/KataGoVisualizer/pull/62/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

ed1d1a8d commented on notebooks/notebooks/iclr2022/estimate-gpu-days-adv.ipynb:
```diff
@@ -0,0 +1,321 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #20.    gpu_type_to_game_throughputs = {}
</pre><p>Could simplify the code below by using a defaultdict(list).</p><p>See https://docs.python.org/3/library/collections.html#defaultdict-examples</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/AlignmentResearch/KataGoVisualizer/pull/62/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

tomtseng replied:
thanks, I didn't know about this class

ed1d1a8d replied:
Learned about it way back when I interned at Dropbox and a senior person told me about it during code review haha

ed1d1a8d reviewed (changes requested):
Just a few small nits regarding changing {} -> defaultdict where appropriate.
Otherwise looks like a very solid estimation.

ed1d1a8d reviewed (approved):
LGTM!

AdamGleave commented:
Remember to rename from iclr2022 before merging!

tomtseng commented:
Rerequesting review because I made several large changes:

* Previously I was only looking at the victimplay GPU usage. I've added the train worker's GPU usage (not counting the time where the train worker is idling and waiting for more rows). The new V100 GPU-days estimate for the cyclic adversary is 2,223 days
* Refactored things into functions so that we can get the passing adversary's training compute (3.07 V100 GPU-days)
* Added a plot of GPU-days vs. win rate for the cyclic-adversary.

The last few points of the plot are:
```
            adv_win     gpu_days
adv_steps                       
434127104  0.000000   634.901149
445174528  0.000000   656.277141
455012352  0.000000   674.590325
464995584  0.120000   693.807642
474976512  0.145833   712.869404
484961536  0.400000   731.345387
494657792  0.645833   749.644650
504708864  0.750000   781.309203
515616512  0.833333   832.093124
525883904  0.875000   931.111608
535865088  0.940000  1206.192888
545065216  0.957224  2223.228956
545993728  0.960000  2883.143384
```

tomtseng commented:
Added plots of GPU-days vs. adversary training steps now (to replace the plots of FLOPs vs. adversary training steps in our paper draft)

ed1d1a8d reviewed (changes requested):
Will take some time to read through the code closer later today. But I was wondering if we could also add the checkpoint switching lines to the win-rate vs compute plot? I'm thinking the lines will spread out towards the end which would be a nice effect to show off clearly.

tomtseng commented:
> Will take some time to read through the code closer later today. But I was wondering if we could also add the checkpoint switching lines to the win-rate vs compute plot? I'm thinking the lines will spread out towards the end which would be a nice effect to show off clearly.

good suggestion, added

tomtseng commented:
Oops, I underestimated the pass adversary GPU-days by a factor of 7 (20.4 V100 GPU-days vs. 3.06 V100 GPU-days) because I forgot that the pass-adversary was run with a single victimplay worker with 7 GPUs rather than 7 victimplay workers with 1 GPU.

Updating paper draft now

AdamGleave commented:
@tomtseng are you still waiting on a review for this, or is it blocked for some other reason?

ed1d1a8d commented:
Ah yeah I wanted to take a look at this but never got around to it cause higher priority things kept popping up. Will try to make some time for this this week.

ed1d1a8d reviewed (approved):
Absolute beast of an analysis notebook. LGTM. I skimmed the code and compared the notebook plots to the FLOPs plots in https://github.com/AlignmentResearch/KataGoVisualizer/blob/main/notebooks/notebooks/paper/estimate-flops-adv.ipynb and figures 5.1 and F.1 in [our paper](https://arxiv.org/pdf/2211.00241.pdf). The numbers seem to check out.
